cuda/solver/common_trs_kernels.cuh

```diff
@@ -389,15 +398,17 @@ is_nan_exact(const T& value)
 }
 
 
-template <bool is_upper, typename ValueType, typename IndexType>
+template <bool is_upper, typename SharedValueType, typename ValueType,
+          typename IndexType>
```

Comment:
Could SharedValueType be deduced inside, instead of making it an additional template parameter? You should be able to pull the code from the kernel launch into here and add a type alias. Otherwise it is easier to accidentally call the kernel with inconsistent types.

Reply:
good idea